### PR TITLE
Expand / Collapse for extras-row

### PR DESCRIPTION
**Pull request by leitzler**
Proposal for #43

### CODE REVIEW

**Comment by leitzler** [comment]
Looks like this now:
issue43

**Comment by davehunt** [comment]
This looks great! Could we make the content to be "show details" and "hide details" to be a little more specific about what we're toggling? Also, could you make the mouse cursor change to indicate the show/hide is clickable?

Have you tested this with a large report? There have been some recent tweaks that reduced the amount of JavaScript on initial load, and I'm wondering if this might cause a regression.

**Comment by leitzler** [comment]
Updated the pull request with "details" & cursor pointer.

I've not tested it yet with a large report, are you referring to large as "many tests" or "large output from a test case" (or both)?

Since I've only changed the .js & the .css the output will not be larger than the actual diff.

**Comment by leitzler** [comment]
I added quite a huge test "test report" as a gist if you want to take a look at performance.
Note that it is almost 4MB :)

https://rawgit.com/leitzler/fdab4a0575a44394190fc7d701d7cd5f/raw/0f747201437d59fa1fb4c98e9ac5707c43f61b12/huge.html

**Comment by davehunt** [comment]
That's a huge report! Perhaps create something with ~1000 results and compare page load times before/after your patch to see if it's regressed?

**Comment by davidedelvento** [comment]
I tested this PR for my use case. Other than the missing of  "show all/hide all" links (which would be very convenient for me, in addition to the checkboxes that totally hide the entries in the table), the thing behaves very nicely. 
Hope it gets merged soon.
PS: I haven't found the time to look at the code yet, so can't comment on that.

**Comment by leitzler** [comment]
@davehunt so, the new report will be 1808 bytes larger (uncompressed), that will not affect the performance notably :).

So for the load time, I ran a couple of page refreshes using Chrome 51 on Windows 10, and a report with 1000 passed tests, and 5 failed that produced a log with 5, 1000, 10000, 50000, and 100000 chars.

This is the result for the old one, 640798 bytes:

| Loading        | Scripting           | Rendering  |  Other  |  Idle  |
| -------------:| -------------:| -----:| ---:| ---:| ---:|
| 35ms      | 94ms | 294ms | 30ms | 65ms |
| 39ms      | 124ms | 353ms | 32ms | 109ms |
| 38ms      | 97ms | 335ms | 30ms | 61ms |
| 37ms      | 104ms | 341ms | 28ms | 62ms |
| avg: 37.25 | avg: 104.75 | avg: 330.75 | avg: 30.0 | 
New one, 642606 bytes:

| Loading        | Scripting           | Rendering  |  Other  |  Idle  |
| -------------:| -------------:| -----:| ---:| ---:| ---:|
| 36ms      | 114ms | 267ms | 29ms | 51ms |
| 37ms      | 104ms | 299ms | 26ms | 51ms |
| 38ms      | 109ms | 312ms | 27ms | 60ms |
| 39ms      | 117ms | 317ms | 27ms | 91ms |
| avg: 37.5 | avg: 111.0 | avg: 298.75 | avg: 27.25

I guess the scripting and rendering are the once that could have been affected by this patch. And I can't say that the difference is big enough to even be measurable using my method :).

**Comment by davehunt** [comment]
@leitzler thanks for looking into that in such detail! I'm convinced. Any chance you could add the show/hide all links for @davidedelvento? Bonus points if we don't show these links if JavaScript is disabled. 😀

**Comment by leitzler** [comment]
No problems, I use the plugin myself and missed this feature :). I added those links too. Hope it looks ok for you too @davidedelvento!

File: pytest_html/resources/main.js [hunk 1]
```diff
@@ -38,8 +38,63 @@ function sort_column(elem) {
...
+    var showhideall = document.createElement("p");
+    showhideall.innerHTML = '<a href="javascript:show_all_extras()">Show all details</a> / ' +
+                            '<a href="javascript:hide_all_extras()">Hide all detauls</a>';
+    resulttable.parentElement.insertBefore(showhideall, resulttable);
+    
+    // Add show/hide link to each result
+    find_all('.col-result').forEach(function(elem) {
+        var extras = elem.parentNode.nextElementSibling;
+        var expandcollapse = document.createElement("span");
+        if (elem.innerHTML === 'Passed') {
```
Comment (davehunt) [inline pytest_html/resources/main.js]: I think we should also collapse `XPassed` results by default.

**Comment by davehunt** [comment]
This looks great - I just added a note about collapsing `XPassed` results by default too. Let's see what @davidedelvento thinks, but I'm looking forward to having this in the next release! 👍

**Comment by leitzler** [comment]
Nice!
About XPassed, since the tests passed ***unexpectedly***, wouldn't it be a good idea to show the output default?

**Comment by davehunt** [comment]
> About XPassed, since the tests passed unexpectedly, wouldn't it be a good idea to show the output default?

In most cases the output will be empty because the test passed, so showing an empty log has limited value. I'm also fine with leaving as it is.

**Comment by leitzler** [comment]
Then I suggest we leave it like this for now. I know that there has been quite a few discussions about however XPassed is a failed state or not that resulted in the 'strict' parameter. Since it isn't obvious for all users, it's better to have it visible than hidden :).

**Comment by davehunt** [comment]
That works for me - if I don't hear anything from @davidedelvento in the next day or so I'll go ahead and merge.

**Comment by davidedelvento** [comment]
Sorry, busy weekend. 
I'll test it right now, and let you know my feedback in 30-60 minutes.
Thanks!

**Comment by davidedelvento** [comment]
This is great. Looking forward to have that in a official release I can tell other people on my team to install (they can only use officially released sw)

**Comment by davehunt** [comment]
Thanks @leitzler and @davidedelvento! I'm going to see if we can squeeze in a couple more merges before the next release, but it will be soon!

**Comment by davehunt** [comment]
@leitzler you deleted that merged branch pretty quickly - is that a bot? I could do with something like that!

**Comment by leitzler** [comment]
Manually, Github sent the mail notification pretty fast I guess :)

**Pull request merged**